### PR TITLE
Use frame umbrella crate in `pallet-proxy` and `pallet-multisig`

### PR DESCRIPTION
**Pull request by kianenigma**
A step towards https://github.com/paritytech/polkadot-sdk/issues/4782

In order to nail down the right preludes in `polkadot-sdk-frame`, we need to migrate a number of pallets to be written with it. Moreover, migrating our pallets to this simpler patter will encourage the ecosystem to also follow along. 

If this PR is approved and has no unwanted negative consequences, I will make a tracking issue to migrate all pallets to this umbrella crate. 

TODO: 

- [x] fix frame benchmarking template. Can we detect the umbrella crate in there and have an `if else`? cc @ggwpez 
- [x] Migrate benchmarking to v2 @re-gius a good candidate for you, you can open a PR against my branch. 
- [x] tracking issue with follow-ups

### CODE REVIEW

File: substrate/frame/multisig/src/weights.rs
Comment (ggwpez) [inline substrate/frame/multisig/src/weights.rs:50]: Huh honestly not sure what to best do here. The long-term solution is to have a project config file for the benchmarker, like for Zepter, so that it gets all the flags from there.  
Then we could add an argument to use the `frame` crate instead. 

Until then we either have to add a `--use-frame-crate` flag manually, or cheat by re-exporting the weight type from the crate and use that.

Comment (kianenigma) [reply substrate/frame/multisig/src/weights.rs:50]: This would break both for `frame` and `polkadot-sdk`, basically meaning pallet development always needs `frame_support` in place? bit unfortunate. 

How about providing two `frame-weight-template.hbs` ? One for old school, one for new school?

Comment (ggwpez) [reply substrate/frame/multisig/src/weights.rs:50]: Discussed about moving all our pallets to include `frame` as dependency and moving the template to assume that exists.

File: substrate/frame/proxy/Cargo.toml
Comment (gui1117) [inline substrate/frame/proxy/Cargo.toml:21]: note `experimental` wasn't used before

File: substrate/frame/src/lib.rs
Comment (gui1117) [inline substrate/frame/src/lib.rs:77]: what is this line?

File: templates/minimal/runtime/Cargo.toml
Comment (gui1117) [inline templates/minimal/runtime/Cargo.toml:41]: personally I find keeping it in one block better.

File: substrate/frame/multisig/src/weights.rs
Comment (gui1117) [reply substrate/frame/multisig/src/weights.rs:50]: maybe we could also assume that the pallet weight is a direct module inside the module with pallet definition.

So we can use `super::pallet::reexport_of_frame`?

But assuming `frame` as a dependency sounds better, people can always modify the `.hbs` to their flavor.

File: substrate/frame/multisig/Cargo.toml
Comment (kianenigma) [reply substrate/frame/multisig/Cargo.toml:21]: I will remove it as soon as I am a bit more confident in the readiness of the `frame` crate 🙈

File: templates/minimal/runtime/Cargo.toml
Comment (shawntabrizi) [reply templates/minimal/runtime/Cargo.toml:41]: then it can also be sorted by toml linters

File: substrate/frame/proxy/src/weights.rs
Comment (shawntabrizi) [inline substrate/frame/proxy/src/weights.rs:49]: this change doesn't make sense. it needs to be done at the template level, or handled later.

File: substrate/frame/multisig/src/weights.rs
Comment (shawntabrizi) [reply substrate/frame/multisig/src/weights.rs:50]: the handlebar templates should be able to allow whatever decision you all pick, but the change needs to happen at that level, not here.

**Review by shawntabrizi — APPROVED** [review]
went through all the files, and changes look good except the benchmarks which need template level changes

File: substrate/frame/multisig/src/weights.rs
Comment (kianenigma) [reply substrate/frame/multisig/src/weights.rs:50]: I see two ways forward with this: 

1. We currently use `substrate/.maintain/frame-weight-template.hbs` in most places. We create a new `substrate/.maintain/frame-umbrella-weight-template.hbs`. Users are free to choose. Our CI should also adapt, and based on the presence of `frame-support` in the deps, make a decision.
2. We make a mega PR that will move all pallets to have a `#[doc(hidden)] pub(crate) mod __private_weights { /* re export all the stuff that might be needed in the weight.rs file */ }`. We also update `substrate/.maintain/frame-weight-template.hbs` to start with `use crate::__private_weights::*`. Our CI should pass once we merge this. Then, we resume this PR. 

Solution 1 is clearly better, and more extensible. Solution 2 is only a reasonable hack if 1 is too hard to pull off. 

@mordamax this is mainly a question from you; how hard will it be to have our weight generation CI (the fellowship's) be able to handle two templates? or perhaps we already support it?

Comment (kianenigma) [reply substrate/frame/multisig/src/weights.rs:50]: I talked to @mordamax and we should explore option, it should not have any major blockers and just needs more updates to our CI.

Comment (kianenigma) [reply substrate/frame/multisig/src/weights.rs:50]: It is not the cleanest thing in the world, but this is how it would look like: 

2d8323f40eb51f81fb66c6a0e969454b16514b37

File: substrate/frame/support/src/lib.rs
Comment (kianenigma) [inline substrate/frame/support/src/lib.rs:889]: I now realize that it is actually best if we don't alter re-exports that are made here, and instead bring more and more things into `frame::prelude`. So I would resume by reverting all changes here, add the items added to `frame::prelude` manually as new `pub use frame_support::{this, that}`. 

This is because adding dependencies here would also mean that we will need to cause a breaking change in frame_support, as we are altering the public API. 

Also, in action we are seeing that it in this PR, adding items in this prelude is causing random other pallets to break, because of ambiguous imports.

Comment (kianenigma) [reply substrate/frame/support/src/lib.rs:889]: So, the guideline that I have written so far: 

> * When doing so, prefer adding the relevant item to `frame_support::pallet_prelude` and `frame_system::pallet_prelude`.

is not a good advice.

File: substrate/frame/multisig/src/weights.rs
Comment (re-gius) [reply substrate/frame/multisig/src/weights.rs:50]: Going with Solution 1: I tested your fix from https://github.com/paritytech/polkadot-sdk/commit/2d8323f40eb51f81fb66c6a0e969454b16514b37 and it works. Also changed the `run_all_benchmarks.sh` script to use both templates in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/49a4cc392aef0a5e2bb5f04b39b4fb17399cf9dd. I don't see any other changes to do.

Comment (re-gius) [reply substrate/frame/multisig/src/weights.rs:50]: Final CI fix in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/21cd6da9b4dfece0f6e84ce99cdad3e3b518ed0a

File: substrate/frame/support/src/lib.rs
Comment (re-gius) [reply substrate/frame/support/src/lib.rs]: edits undone in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/bd17cf205f6ddc934436eb2a3d68d63fb7e86b02

File: templates/minimal/runtime/Cargo.toml
Comment (re-gius) [reply templates/minimal/runtime/Cargo.toml]: reverted in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/bd17cf205f6ddc934436eb2a3d68d63fb7e86b02

File: .github/scripts/cmd/cmd.py
Comment (ggwpez) [inline .github/scripts/cmd/cmd.py:179]: https://github.com/paritytech/polkadot-sdk/issues/6204

File: substrate/frame/multisig/src/benchmarking.rs
Comment (ggwpez) [inline substrate/frame/multisig/src/benchmarking.rs:72]:
```suggestion
	/// `s`: Signatories, need at least 2 people
```
It should also be possible to put doc comments here.

File: substrate/frame/src/lib.rs
Comment (ggwpez) [inline substrate/frame/src/lib.rs:234]: host functions?

File: substrate/frame/src/lib.rs
Comment (ggwpez) [inline substrate/frame/src/lib.rs:241]:
```suggestion
	#[deprecated(note = "'The V1 benchmarking syntax is deprecated. Please use the V2 syntax. This warning may become a hard error any time after April 2025. For more info, see: https://github.com/paritytech/polkadot-sdk/pull/5995")]
```

File: substrate/frame/src/lib.rs
Comment (ggwpez) [inline substrate/frame/src/lib.rs:299]: Should also mention when this will finally be removed. Otherwise people never switch 😆

File: substrate/scripts/run_all_benchmarks.sh
Comment (ggwpez) [inline substrate/scripts/run_all_benchmarks.sh:117]: What is happening here? I am generally not a fan of formatting changes when they are not enforced by CI.  
One dev changes it to a specific way and the next dev changes it back.

It also makes it much slower to review, since i now would need to check every line that was touched.

**Review by ggwpez — COMMENTED** [review]
Looks good, just some comments.

File: substrate/frame/multisig/src/benchmarking.rs
Comment (re-gius) [reply substrate/frame/multisig/src/benchmarking.rs]: Changed all to doc comments in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/d46029526908a15a9d76c738afc92168643cea1b

File: substrate/frame/democracy/src/migrations/v1.rs
Comment (re-gius) [reply substrate/frame/democracy/src/migrations/v1.rs]: True, because we removed `Bounded` from `frame_support::pallet_prelude::*`. Reverted in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/d46029526908a15a9d76c738afc92168643cea1b

File: substrate/scripts/run_all_benchmarks.sh
Comment (re-gius) [reply substrate/scripts/run_all_benchmarks.sh]: I see, went back to previous formatting in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/d46029526908a15a9d76c738afc92168643cea1b

File: substrate/frame/src/lib.rs
Comment (kianenigma) [reply substrate/frame/src/lib.rs:301]: Good point! Typically we allocate 6-9 months? So you can add a note that it will be removed June 2025.

File: .github/scripts/cmd/cmd.py
Comment (kianenigma) [inline .github/scripts/cmd/cmd.py:263]: @mordamax can you review this file please?

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs
Comment (kianenigma) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs:62]: This may have been due to my previous work.. Can you check if it can be reverted now?

File: prdoc/pr_5995.prdoc
Comment (kianenigma) [reply prdoc/pr_5995.prdoc:9]: noting this because we have two umbrella crates: 
- polkadot-sdk
- polkadot-sdk-frame

File: substrate/frame/multisig/Cargo.toml
Comment (kianenigma) [reply substrate/frame/multisig/Cargo.toml:21]: What we should try here is to see if removing the `feature = runtime` still works? 

The idea in the `frame` umbrella crate was that for pallet development, you would only need crate itself, and for a runtime, the crate plus `runtime` feature.

File: substrate/frame/src/lib.rs
Comment (kianenigma) [inline substrate/frame/src/lib.rs:262]:
```suggestion
```

File: prdoc/pr_5995.prdoc
Comment (kianenigma) [inline prdoc/pr_5995.prdoc:18]:
```suggestion
  - name: polkadot-sdk-frame
```

?

File: prdoc/pr_5995.prdoc
Comment (kianenigma) [inline prdoc/pr_5995.prdoc:19]:
```suggestion
    bump: major
```

File: prdoc/pr_5995.prdoc
Comment (kianenigma) [inline prdoc/pr_5995.prdoc:15]: This seems correct to me -- the interface of these two pallets is the same, but altering deps is a backwards compatible new feature, and should be minor, not patch. 

Can you double check this please @ggwpez?

File: substrate/frame/multisig/src/tests.rs
Comment (kianenigma) [inline substrate/frame/multisig/src/tests.rs:24]: One little detail here is which preludes are a superset of the others. 

I recall I made a deliberate choice about preludes being a superset of one another, but I didn't update the code 🙈 For example here, I think `testing_prelude` brings in the other two

File: substrate/frame/src/lib.rs
Comment (kianenigma) [inline substrate/frame/src/lib.rs:37]: Can you add a note here about the fact that some preludes already include one another: 

The order is: 

`testing` brings in `frame::prelude::*` and `runtime::prelude`
`runtime` brings in `frame::predlude::*` 
`benchmarking` brings in `frame::predlude::*` 
 etc etc

File: templates/minimal/pallets/template/src/lib.rs
Comment (kianenigma) [reply templates/minimal/pallets/template/src/lib.rs:9]: Good point, I was trying to keep this _minimal_. 

@re-gius can you add one `u32` storage value in this crate to fix this? then `#[allow(unused_imports)]` can be removed.

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs
Comment (re-gius) [reply bridges/snowbridge/pallets/inbound-queue/src/lib.rs]: reverted in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/88265e59c283c83335e81f8115378efe174fb9a1

File: prdoc/pr_5995.prdoc
Comment (ggwpez) [reply prdoc/pr_5995.prdoc:15]: Yea, in these simple cases the CI should also fail if its incorrectly attributed a patch. So just try patch.

File: templates/minimal/runtime/Cargo.toml
Comment (kianenigma) [inline templates/minimal/runtime/Cargo.toml:16]: it works without this? 

Need to figure out the relationship between `polkadot-sdk` parent umbrella, and how it can be used to pull in `polkadot-sdk-frame` only. I thought `experimental` was needed here.

File: substrate/frame/src/lib.rs
Comment (re-gius) [reply substrate/frame/src/lib.rs:37]: Added in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/dc2973d0074a6a41f840b042cb204f310cb9a04d

File: substrate/frame/src/lib.rs
Comment (re-gius) [reply substrate/frame/src/lib.rs:41]: Added in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/dc2973d0074a6a41f840b042cb204f310cb9a04d

File: substrate/frame/democracy/src/migrations/v1.rs
Comment (kianenigma) [reply substrate/frame/democracy/src/migrations/v1.rs]: This is a good example of where adding more things to frame_support::pallet_prelude::* turned out to be a headache.

File: substrate/frame/src/lib.rs
Comment (kianenigma) [reply substrate/frame/src/lib.rs:41]: quick note, I think your comment can be more clear as to: 

1. If you want to build a pallet with this, import it as:


```
polkadot-sdk-frame = { version = "foo" }
```

If you want to build a runtime, import it as

```
polkadot-sdk-frame = { version = "foo", features = ["runtime"] }
```

**Comment by kianenigma** [comment]
Feel free to approve this on my behalf once all open converstaions are serviced one last time :)

File: substrate/frame/src/lib.rs
Comment (re-gius) [reply substrate/frame/src/lib.rs:41]: Thanks, changed in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/48050fd5c66c3c587c4fc7ecaa1fdbf1b5670827

File: templates/minimal/pallets/template/src/lib.rs
Comment (re-gius) [reply templates/minimal/pallets/template/src/lib.rs]: Added in https://github.com/paritytech/polkadot-sdk/pull/5995/commits/5918bef5c71113c5bf8b3c7b135ea8cf50672953

File: templates/minimal/runtime/Cargo.toml
Comment (re-gius) [reply templates/minimal/runtime/Cargo.toml:16]: `build` works, I don't know if there's anything else to check

**Review by re-gius — APPROVED** [review]
Approving on behalf of @kianenigma

File: substrate/frame/src/lib.rs
Comment (kianenigma) [inline substrate/frame/src/lib.rs:89]:
```suggestion
//! ```text
```

File: substrate/frame/src/lib.rs
Comment (kianenigma) [inline substrate/frame/src/lib.rs:83]:
```suggestion
//! ```text
```

**Comment by kianenigma** [comment]
@pepoviola similar flaky ZN tests failing here.

**Review by ggwpez — APPROVED** [review]
Approving for Guiseppe.

**Comment by re-gius** [comment]
Solves https://github.com/paritytech/polkadot-sdk/issues/6160

**Pull request merged**
